src/Evolution/SlopeLimiters/Minmod.hpp

```diff
@@ -0,0 +1,413 @@
+#include "ErrorHandling/Assert.hpp"
+#include "NumericalAlgorithms/LinearOperators/Linearize.hpp"
+#include "NumericalAlgorithms/LinearOperators/MeanValue.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeArray.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+
+namespace SlopeLimiters {
+template <size_t VolumeDim, typename TagsToLimit>
+class Minmod;
+}  // namespace SlopeLimiters
+
+template <size_t>
+class Mesh;
+/// \endcond
+
+/// \ingroup SlopeLimitersGroup
+/// \brief Things relating to slope limiting.
+namespace SlopeLimiters {
+/// \ingroup SlopeLimitersGroup
+/// \brief Possible types of the minmod slope limiter.
+enum class MinmodType { LambdaPi1, LambdaPiN, Muscl };
+}  // namespace SlopeLimiters
+
+namespace Minmod_detail {
+// Encodes the return status of the minmod_tvbm function.
+struct MinmodResult {
+  const double value;
+  const bool activated;
+};
+// The TVBM-corrected minmod function, see e.g. Cockburn reference Eq. 2.26.
+MinmodResult minmod_tvbm(double a, double b, double c,
+                         double tvbm_scale) noexcept;
+
+template <typename TensorType>
+using tensor_double_from = Tensor<double, typename TensorType::symmetry,
+                                  typename TensorType::index_list>;
+
+// Implements the minmod limiter for one Tensor<DataVector> at a time.
+template <size_t VolumeDim, typename Tag>
+bool limit_one_tensor(
+    const gsl::not_null<std::add_pointer_t<db::item_type<Tag>>>& local_tensor,
+    const std::unordered_map<Direction<VolumeDim>,
+                             tensor_double_from<db::item_type<Tag>>>&
+        neighbor_tensors,
+    const SlopeLimiters::MinmodType& type, const double tvbm_constant,
+    const Element<VolumeDim>& element, const Mesh<VolumeDim>& mesh,
+    const tnsr::I<DataVector, VolumeDim, Frame::Logical>& logical_coords,
+    const tnsr::I<double, VolumeDim>& element_size,
+    const std::unordered_map<Direction<VolumeDim>, tnsr::I<double, VolumeDim>>&
+        neighbor_sizes) noexcept {
+  // True if the mesh is linear-order in every direction
+  const bool mesh_is_linear = (mesh.extents() == Index<VolumeDim>(2));
+
+  const double tvbm_scale = [&tvbm_constant, &element_size]() {
+    double max_h_sqr = 0.0;
+    for (size_t d = 0; d < VolumeDim; ++d) {
+      max_h_sqr = std::max(max_h_sqr, square(element_size.get(d)));
+    }
+    return tvbm_constant * max_h_sqr;
+  }();
+
+  // Results from SpECTRE paper (https://arxiv.org/abs/1609.00098) used a
+  // max_slope_factor a factor of 2.0 too small, so that LambdaPi1 behaved
+  // like MUSCL, and MUSCL was even more dissipative.
+  const double max_slope_factor =
+      (type == SlopeLimiters::MinmodType::Muscl) ? 1.0 : 2.0;
+
+  bool some_component_was_limited = false;
+  for (size_t i = 0; i < local_tensor->size(); ++i) {
```

Comment:
If you mean this loop cannot be factored because of tensor-ness: is it possible to factor if you use iterators? I believe all Tensors have the same iterator type for this exact reason. I suspect the resulting code won't be super beautiful, but it would be nice to have it in only a single translation unit.

Reply:
Nils and I discussed hiding the tensor type in `apply`, and calling `limit_one_tensor` without any type information. Then, `limit_one_tensor` can be moved to the source file.

Nils suggested creating a (in `apply`)
`no_type_tensor = std::unordered_map<Direction<VolumeDim>, "tensor begin iterator">{...}`

so that code that previously used (in `limit_one_tensor`)
`blah = neighbor_tensors.at(dim)[i]`

would become
`blah = *(no_type_tensor.at(dim) + std::distance(local_iter, tensor_begin_iter))`

I will try this.